Feature/add node api

seawatts opened the pull request.

Commit:
Updated readme with steps to build project

Commit:
Checkpoint: Ability to POST a node
Checkpoint: Ability to GET a list of nodes
Checkpoint: Creating tests around these two methods

Commit:
Refactored tests in node handler
Create nodes in node handler takes a list now

Commit:
Fixed compilation errors after rebase

Commit:
Fixed tests not passing because of node codec not including provisionerResults

albertshau commented on README.md:
```diff
@@ -18,6 +18,47 @@ To learn more about Continuuity Loom, here are the resources available:
       * [Documentation 0.9.6](http://www.continuuity.com/docs/loom/0.9.6/en/index.html)
       * [Documentation 0.9.5](http://www.continuuity.com/docs/loom/0.9.5/en/index.html)
 
+## Building Source
```
nice, thanks for adding this

seawatts replied:
No problem. Not fully finished though i guess didnt put in the Running source part or Running in IntelliJ

albertshau commented on README.md:
```diff
@@ -18,6 +18,47 @@ To learn more about Continuuity Loom, here are the resources available:
+## Building Source
+
+**Backend Java Server**
+_Prerequisites_
```
when you view the rendered file, is prerequisites showing up on the same line as "backend java server" for you?

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,155 @@
+@Path("/v1/loom/nodes")
+public class LoomNodeHandler extends LoomAuthHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(LoomNodeHandler.class);
+  //  private final NodeService clusterService;
```
can remove this if its not needed

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,155 @@
+   * Initializes a new instance of a LoomNodeHandler.
+   */
+  @Inject
+  public LoomNodeHandler(TenantStore tenantStore, NodeService nodeService, NodeStoreService nodeStoreService,
```
It's better if the constructor is private to prevent people from trying to new it.

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  /**
+   * Get all noes visible to the user.
```
'nodes' typo

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,155 @@
+    try {
+      nodes = nodeStoreService.getView(account).getAllNodes();
+    } catch (IOException e) {
+      responder.sendError(HttpResponseStatus.INTERNAL_SERVER_ERROR, "Exception getting nodes.");
```
we can return here to prevent null pointers further on

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,155 @@
+      JsonObject obj = new JsonObject();
+      obj.addProperty("id", node.getId());
+      obj.addProperty("clusterId", node.getClusterId());
+      //      obj.add("properties", node.getProperties());
```
we don't want properties here?

seawatts replied:
Oops forgot to add those

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,155 @@
+    if (account == null) {
+      return;
+    }
+    Reader reader = new InputStreamReader(new ChannelBufferInputStream(request.getContent()), Charsets.UTF_8);
```
should close the reader

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,155 @@
+      responder.sendError(HttpResponseStatus.INTERNAL_SERVER_ERROR, "Exception creating nodes.");
+    }
+
+    responder.sendStatus(HttpResponseStatus.CREATED);
```
should move this into the try block. Otherwise, if there's an error writing we'll write to the responder once for the error and then again here (which will result in an exception)

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,155 @@
+    try {
+      nodeStoreService.getView(account).writeNodes(nodes);
+    } catch (IllegalAccessException e) {
+      responder.sendError(HttpResponseStatus.INTERNAL_SERVER_ERROR, "Exception creating nodes.");
```
would be more accurate to return a forbidden status since this exception indicates the account cant access the resource

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,155 @@
+
+    responder.sendStatus(HttpResponseStatus.NO_CONTENT);
+  }
+}
```
should there be an endpoint to get a specific node?

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,155 @@
+    responder.sendStatus(HttpResponseStatus.CREATED);
+  }
+
+  @PUT
```
the PUT should contain the node id in the path, since we're writing a single nodes and not the collection of nodes.

albertshau commented on server/src/test/java/com/continuuity/loom/http/LoomNodeHandlerTest.java:
```diff
@@ -0,0 +1,132 @@
+    String nodeJsonString = getNodeAsJsonString(node);
+    HttpResponse response = doPut("/v1/loom/nodes", nodeJsonString, USER1_HEADERS);
+    assertResponseStatus(response, HttpResponseStatus.NO_CONTENT);
+    // test to see that it has been updated
```
can you add this part

Commit:
Update README.md

Commit:
Fixed: Update node test not actually checking to see if the node had been updated
Added: Get single node in api
Added: Node id required when updating a node
Added: Returning all the fields in the node when getting a node from the api

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,199 @@
+      return;
+    }
+    Reader reader = new InputStreamReader(new ChannelBufferInputStream(request.getContent()), Charsets.UTF_8);
+    Node node = gson.fromJson(reader, Node.class);
```
should check if the node we're given in the body has the same id as whats in the path and 400 if they're different.

albertshau replied:
should 400 if there is any other exception reading the body too.

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    JsonArray jsonArray = new JsonArray();
+    for (Node node : nodes) {
```
actually, I think it will work if you send the set directly in sendJson

albertshau commented:
couple more small comments, otherwise lgtm

Commit:
Added support to throw 400 errors when node ID does not match the URI
Refactored json response back to the client

seawatts commented:
Alright, fixed those last couple comments

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,193 @@
+
+    try {
+      Node node = gson.fromJson(reader, Node.class);
+      if (node.getId().equals(nodeId)) {
```
Since nodes can be input by users, we'll need to check that the id is not null.  I think it makes sense to put the check in the node constructor since id shouldn't ever be missing.

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,193 @@
+      return;
+    }
+    Reader reader = new InputStreamReader(new ChannelBufferInputStream(request.getContent()), Charsets.UTF_8);
+    Set<Node> nodes = gson.fromJson(reader, new TypeToken<Set<Node>>() {}.getType());
```
I think it's simpler to only accept one node at a time instead of multiple.  Otherwise we could have situations where some node in the middle of the set is invalid (maybe the id already exists or something like that).  Or somebody posts in a humongous set and the server OOMs.

albertshau replied:
here we'll also have to check that the body is valid json and 400 if not.

albertshau commented:
couple more comments, then lgtm.

albertshau commented on server/src/main/java/com/continuuity/loom/http/handler/LoomNodeHandler.java:
```diff
@@ -0,0 +1,193 @@
+    } catch (JsonIOException e){
+      responder.sendError(HttpResponseStatus.BAD_REQUEST, "Exception reading node from body.");
+    } catch (JsonSyntaxException e){
+        responder.sendError(HttpResponseStatus.BAD_REQUEST, "Exception reading node from body.");
```
indentation

Commit:
Can only post one node at a time now
Added extra error handling around adding and updating nodes
Fixed code style

seawatts commented:
Alright fixed comments

albertshau commented:
:+1:
